presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1102,27 +1105,28 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                if (!isSupportedJoinConjunct(conjunct)) {
+                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported expression in join clause: %s", conjunct);
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
```

Comment:
@cberner. Thanks for comments.
I could use some help here :)

I tried to refactor the code into direction suggested by you but it have problems making it work. Either I misunderstood you or have to little understanding in how semantic analyzer/relation planner work (or both :) ).

I made two approaches:

1. I skipped the `DependencyExtractor` part in `SemanticAnalyzer` and `RelationPlanner`. Just performed analysis of all join condition conjuncts in context of `output` relation. Semantic analysis was passing but I got errors in planner.

I got error when (in case of equality conjunct) expression for left side expression was being added to plan for left join side. The code blow failed in `rewriteBase.getSymbol` as `analysis.getFieldIndex(node)` returned wrong index. And wrong index was stored in analysis object as analysis of join condition was done in context of `output` relation which different fields than `left` relation.

This is the part of `TranslationMap.rewriteExpressionWithResolvedName` which ultimately fails:
```java
       private Expression rewriteExpressionWithResolvedName(Expression node)
       {
           Optional<Integer> fieldIndex = analysis.getFieldIndex(node);
           checkState(fieldIndex.isPresent(), "No field mapping for node '%s'", node);

           Symbol symbol = rewriteBase.getSymbol(fieldIndex.get());
```

Sandbox code for this approach is here https://github.com/losipiuk/presto/tree/non-equi-join-sandbox.

2. In second approach I changed `ExpressionAnalyzer` in such way that we could pass more than one `RelationType` as `tupleDescriptor` parameter to `analyze` method.
Then in places where `tupleDescriptor.indexOf` and `tupleDescriptor.resolveFields` was called, I iterated over passed `tupleDescriptors` and returned result from one which could provide it. There was assumption that field sets in provided list of tupleDiscriptors was disjoint so only one should return result.

With this change planning finished but I got errors from `PlanSanityChecker` 

For query:
```
SELECT COUNT(clerk) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey APPROXIMATE AT 99.999 CONFIDENCE"
```
I got following error:
```
java.lang.IllegalArgumentException: Invalid node. Predicate dependencies ([lineitem, orders]) not in source plan output ([orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, shipdate, commitdate, receiptdate, shipinstruct, shipmode, comment, row_number, $sampleweight, expr, orderkey_0, custkey, orderstatus, totalprice, orderdate, orderpriority, clerk, shippriority, comment_1, row_number_2, $sampleweight_3, expr_4])
```

Not yet sure why this is happening. 
Sandbox code for this approach is here: https://github.com/losipiuk/presto/tree/non-equi-join-sandbox2


I would be super helpful for any hints on how to tackle the problem. Do you still think that analysing join expression in context of output relation is right choice? It seem proper to me but I am not sure how vast changes such approach require. Nevertheless it will probably be needed anyway, if we want to support complex expressions, other that inequalities with one side of inequality only touching columns from one side of join, as join condition.


One more thing concerning your comment. Could you elaborate on this sentence:
```
Then I think you'll be able to remove all this code with dependency extraction, and split the expressions into left and right
```

How do you envision splitting into left and right subexpressions of join condition for such query
```
select * from a join b on f(a.x, b.y) = 1
``` 
? what would be the result of splitting?
Are there any tools in the codebase already to do such splitting?